### PR TITLE
Adopt backward iteration (KIP-617) on Window Stores to be available on kafka-streams-2.7

### PR DESCRIPTION
**Pull request by jeqo**
[KIP-617](https://cwiki.apache.org/confluence/display/KAFKA/KIP-617%3A+Allow+Kafka+Streams+State+Stores+to+be+iterated+backwards) introduces backward iteration on all stores. 

This changes include:
- Migrating from key-value stores into window stores: by using windows, we remove the need to track traces->timestamp ourselves, and can use windowed keys that do this already.
- Reverse iteration solves the issue to iterate windows from oldest to newest, adding complexity to return latest results that match a query.
- Another side effect of using WindowStores instead of KeyValues is that we can use Window retention (remove old windows) to be handled by the library it self, instead of managing kv retention ourselves.

Testing:
While we wait for kafka2.7 to be release, we can test against its snapshot:

```
git clone git@github.com:apache/kafka
git checkout 2.7
./gradlewAll install
```

### CODE REVIEW

File: storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java [hunk 1]
```diff
@@ -144,72 +139,67 @@
...
+    WindowStore<String, Set<String>> tagsStore = testDriver.getWindowStore(AUTOCOMPLETE_TAGS_STORE_NAME);
+    Map<String, Set<String>> tags = new HashMap<>();
+    try (
+      final KeyValueIterator<Windowed<String>, Set<String>> fetch = tagsStore.fetchAll(0, 10010L)) {
+      while (fetch.hasNext()) {
+        final KeyValue<Windowed<String>, Set<String>> next = fetch.next();
+        tags.put(next.key.key(), next.value);
+      }
+    }
+    assertThat(tags.get("environment")).containsExactly("dev");
```
Comment (adriancole) [inline storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java:202]: might be a way in assertj iterator assert to do similar.. something like 
```java
.extracting(next -> next.key.key(), next -> next.value)
.containsExactly(tuple("environment", collections.singleton("dev") 
```

**Review by adriancole — APPROVED** [review]
Looks cool. I suppose when this is in, we can try to start weaving in the normal IT tests, maybe starting with auto-complete?

**Review by jeqo — COMMENTED** [review]
added steps to test it locally on the description.

File: storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java (hunk 1 above)
Comment (jeqo) [reply storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java]: give a try. lmk if it looks good

File: storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java [hunk 2]
```diff
@@ -144,72 +140,70 @@
...
-    // Then: Traces store is empty
-    assertThat(traces.get(a.traceId())).isNull();
+    WindowStore<String, List<Span>> traces = testDriver.getWindowStore(TRACES_STORE_NAME);
+    try (final WindowStoreIterator<List<Span>> fetch = traces.fetch(a.traceId(), 0, 10000L)) {
+      assertThat(fetch.hasNext()).isTrue();
+      final KeyValue<Long, List<Span>> next = fetch.next();
+      assertThat(next.value).isEqualTo(spans);
+    }
+    WindowStore<String, Set<String>> spanNames = testDriver.getWindowStore(SPAN_NAMES_STORE_NAME);
+    Map<String, Set<String>> serviceNames = new HashMap<>();
```
Comment (adriancole) [inline storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java:182]: by mentioning using `.extracting` I meant to see if you can avoid copying into a map first. ex assert on the WindowStore.fetchAll

not a biggie and I can give a try if you like (tomorrow)

Comment (jeqo) [reply storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java]: Right! Got it. Hope it looks better now 😅

File: storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java [hunk 3]
```diff
@@ -144,72 +137,68 @@
...
+      assertThat(fetch).isExhausted();
+    }
+    WindowStore<String, Set<String>> spanNames = testDriver.getWindowStore(SPAN_NAMES_STORE_NAME);
+    try (
+      final KeyValueIterator<Windowed<String>, Set<String>> fetch = spanNames.fetchAll(0, 10000L)) {
+      assertThat(fetch).hasNext();
+      assertThat(fetch).hasNext();
+      assertThat(fetch.next())
+        .extracting(next -> next.key.key(), next -> next.value)
+        .containsExactly("svc_a", Collections.singleton("op_a"));
```
Comment (adriancole) [inline storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java:187]: better!

File: storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java [hunk 4]
```diff
@@ -144,72 +137,68 @@
...
+      assertThat(fetch.next())
+        .extracting(next -> next.value)
+        .isEqualTo(spans);
+      assertThat(fetch).isExhausted();
+    }
+    WindowStore<String, Set<String>> spanNames = testDriver.getWindowStore(SPAN_NAMES_STORE_NAME);
+    try (
+      final KeyValueIterator<Windowed<String>, Set<String>> fetch = spanNames.fetchAll(0, 10000L)) {
+      assertThat(fetch).hasNext();
+      assertThat(fetch).hasNext();
```
Comment (adriancole) [inline storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java:184]: just in case? :p

Comment (jeqo) [reply storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java]: 😂 good catch

**Comment by adriancole** [comment]
ps if you merge this, I'll update the build to latest hotness

**Comment by jeqo** [comment]
2.7 is released. I feel this is ready to be merged.

**Pull request merged**